src/audio/kpb.c

```diff
@@ -1742,13 +1742,22 @@ static enum task_state kpb_draining_task(void *arg)
 	uint64_t current_time;
 	size_t period_bytes = 0;
 	size_t period_bytes_limit = draining_data->pb_limit;
-	size_t period_copy_start = sof_cycle_get_64();
+	size_t period_copy_start;
 	size_t time_taken;
 	size_t *rt_stream_update = &draining_data->buffered_while_draining;
 	struct comp_data *kpb = comp_get_drvdata(draining_data->dev);
 	bool sync_mode_on = draining_data->sync_mode_on;
 	bool pm_is_active;
 
+	/*
+	 * WORKAROUND: The code below accesses KPB sink buffer and calls comp_copy() on
+	 * component connected to sink. EDF task thread has preemptible low priority and
+	 * so can be preempted by LL thread. This could result in broken state of sink buffer
+	 * or component connected to sink.
+	 * Hence k_sched_lock() is used temporary to block LL from preempting EDF task thread.
+	 */
+	k_sched_lock();
+
```

Comment:
doesn't this break the concept of LL? The draining part could take quite a while for a large buffer, and even with this cycle of lock/unlock the entire premise of LL executing within 1ms is broken.

That's a priority inversion in my book where a low-priority task prevents a high-priority/low-latency one from being executed...

Why can't we fix the broken state of the sink instead ?

Reply:
On each loop iteration draining task basically just calls comp_copy() on component connected to KPB sink. So each loop iteration should take much less then 1 ms.
If timer interrupt increased semaphore on which LL thread waits, immediately at k_yield() scheduler will switch to executing LL thread since LL thread has higher priority then EDF task thread. Or am I overlooking something?

Looks like original draining task code was written in pre-Zephyr era. And probably the code works or worked without Zephyr just fine. However, the code seems were never properly ported to Zephyr. Original draining task implementation has multiple problems:
* there is no sleep/yield in loop iteration and so task blocks IPC thread for a long time (several seconds in our test) and waists power (CPU cannot reach idle between loop iterations).
* LL can preempt draining task and call comp_copy() for same component or access/modify same buffer/stream.

This fix is an attempt to fix CI failure, not to do proper reimplementation of draining task. Without the fix we have random failures on KPB test. We have to remove test temporary just not to block all the work on SOF.

The test fails with draining task runs forever because KPB sink audio_stream has broken state: avail == free == 0. Such state normally cannot happen as those values are always calculated as avail = size - free and free = size - avail. avail and free simultaneously equal to 0 can only happen when LL preempts draining task in unlucky moment.

The fix passes the test. It has no intent to re-implement draining task properly: I'm not sure if KPB is used in any production case other then demo. The intent is to fix CI and return back KPB test.

Reply:
Like you said no one is using KPB in any firmware based on Zephyr (post 2.2).

I am not getting the idea of 'fixing CI' which consists in re-testing a broken solution based on a workaround impacting other usages.

Is KPB is so badly broken, it needs to be disabled and redone, no?

Reply:
I think there's value in keeping kpb tests. This exercizes a certain type of flow (running an EDF task) in CI that we don't have many others in CI and if we remove the test, this code will bitrot very quickly. The workaround here is local to kpb, so I see no harm outside kpb.

Reply:
@kv2019i "The workaround here is local to kpb" -> not really, it would impact any other compound use case with e.g. speaker protection running.

If KPB is a test component, them let's mark it as non-default and opt-in for CI purposes. This shouldn't go in any releases IMHO.